ENH: Vectorize quicksort for 16-bit and 64-bit dtype using AVX512

r-devulap opened the pull request:
This patch adds AVX512 based 64-bit on AVX512-SKX and 16-bit sorting on AVX512-ICL. All the AVX512 sorting code has been reformatted as a separate header files and put in a separate folder. The AVX512 64-bit sorting is nearly 10x faster and AVX512 16-bit sorting is nearly 16x faster when compared to `std::sort`. ~~Still working on running NumPy benchmarks to get exact benchmark numbers~~

16-bit int sped up by **17x** and float64 by nearly **10x** for random arrays. Benchmarked on a 11th Gen Tigerlake i7-1165G7. 

```
      before           after         ratio
     [41b6ac0e]       [2b384ac6]
     <main>           <avxsort> 
-        44.2±1μs       41.8±0.5μs     0.95  bench_function_base.Sort.time_sort('quick', 'float32', ('sorted_block', 100))
-      39.7±0.1μs      37.1±0.02μs     0.93  bench_function_base.Sort.time_sort('quick', 'int32', ('sorted_block', 1000))
-        45.7±3μs       42.6±0.4μs     0.93  bench_function_base.Sort.time_sort('quick', 'float32', ('random',))
-      39.8±0.1μs       37.1±0.4μs     0.93  bench_function_base.Sort.time_sort('quick', 'int32', ('random',))
-     39.1±0.03μs      36.4±0.03μs     0.93  bench_function_base.Sort.time_sort('quick', 'int32', ('sorted_block', 100))
-      39.9±0.1μs       37.1±0.2μs     0.93  bench_function_base.Sort.time_sort('quick', 'int32', ('reversed',))
-      39.4±0.2μs       36.3±0.2μs     0.92  bench_function_base.Sort.time_sort('quick', 'int32', ('ordered',))
-     40.6±0.03μs       37.1±0.3μs     0.91  bench_function_base.Sort.time_sort('quick', 'uint32', ('sorted_block', 100))
-      46.4±0.7μs       41.8±0.4μs     0.90  bench_function_base.Sort.time_sort('quick', 'float32', ('ordered',))
-      40.5±0.4μs       36.4±0.1μs     0.90  bench_function_base.Sort.time_sort('quick', 'int32', ('sorted_block', 10))
-        42.5±1μs      37.5±0.08μs     0.88  bench_function_base.Sort.time_sort('quick', 'uint32', ('sorted_block', 1000))
-        42.4±1μs       36.8±0.4μs     0.87  bench_function_base.Sort.time_sort('quick', 'uint32', ('ordered',))
-        43.4±3μs       37.4±0.2μs     0.86  bench_function_base.Sort.time_sort('quick', 'uint32', ('sorted_block', 10))
-        44.5±1μs         38.0±1μs     0.85  bench_function_base.Sort.time_sort('quick', 'uint32', ('reversed',))
-        81.7±4μs       67.9±0.2μs     0.83  bench_function_base.Sort.time_sort('quick', 'float64', ('ordered',))
-        45.7±1μs       37.6±0.2μs     0.82  bench_function_base.Sort.time_sort('quick', 'uint32', ('random',))
-       119±0.6μs       77.7±0.2μs     0.65  bench_function_base.Sort.time_sort('quick', 'int64', ('reversed',))
-         136±5μs       67.1±0.3μs     0.49  bench_function_base.Sort.time_sort('quick', 'float64', ('reversed',))
-     70.6±0.03μs      33.9±0.07μs     0.48  bench_function_base.Sort.time_sort('quick', 'int16', ('ordered',))
-         113±1μs       34.8±0.1μs     0.31  bench_function_base.Sort.time_sort('quick', 'int16', ('reversed',))
-         325±9μs      80.9±0.07μs     0.25  bench_function_base.Sort.time_sort('quick', 'int64', ('sorted_block', 1000))
-         452±8μs         85.7±4μs     0.19  bench_function_base.Sort.time_sort('quick', 'int64', ('sorted_block', 100))
-         386±5μs       70.3±0.6μs     0.18  bench_function_base.Sort.time_sort('quick', 'float64', ('sorted_block', 1000))
-         460±5μs         82.0±3μs     0.18  bench_function_base.Sort.time_sort('quick', 'int64', ('sorted_block', 10))
-      94.5±0.5ms       14.2±0.4ms     0.15  bench_function_base.Sort.time_sort_worst
-        537±10μs         80.8±2μs     0.15  bench_function_base.Sort.time_sort('quick', 'int64', ('random',))
-        518±10μs         73.9±3μs     0.14  bench_function_base.Sort.time_sort('quick', 'float64', ('sorted_block', 100))
-        533±20μs       74.5±0.3μs     0.14  bench_function_base.Sort.time_sort('quick', 'float64', ('sorted_block', 10))
-        70.3±1μs       8.54±0.2μs     0.12  bench_function_base.Sort.time_sort('quick', 'int64', ('uniform',))
-        652±30μs         70.6±1μs     0.11  bench_function_base.Sort.time_sort('quick', 'float64', ('random',))
-         317±3μs       32.5±0.2μs     0.10  bench_function_base.Sort.time_sort('quick', 'int16', ('sorted_block', 1000))
-         439±2μs         34.4±2μs     0.08  bench_function_base.Sort.time_sort('quick', 'int16', ('sorted_block', 100))
-         133±6μs       9.99±0.4μs     0.07  bench_function_base.Sort.time_sort('quick', 'float64', ('uniform',))
-         452±1μs       33.4±0.1μs     0.07  bench_function_base.Sort.time_sort('quick', 'int16', ('sorted_block', 10))
-      63.5±0.8μs       4.44±0.2μs     0.07  bench_function_base.Sort.time_sort('quick', 'int16', ('uniform',))
-         555±7μs       32.8±0.1μs     0.06  bench_function_base.Sort.time_sort('quick', 'int16', ('random',))

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
```


<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

Commit:
ENH: Add x86-simd-sort source files

r-devulap commented:
@serge-sans-paille: Reformatted your C++ code and added 16-bit and 64-bit sorting.

r-devulap commented:
Added benchmark numbers. Need to fix the CI errors.

Commit:
ENH: Add AVX-512 based 64-bit dtype sort

Commit:
ENH: Add AVX-512 based 16-bit dtype sort

Commit:
BUG: Use longlong when NPY_SIZEOF_LONG is 4

Commit:
Revert "ENH: Add AVX-512 based 16-bit dtype sort"

This reverts commit 225c8bab83d239d8888bc7b688efed97ab2284cf.

mattip commented:
The [cygwin failure](https://github.com/numpy/numpy/actions/runs/3148780125/jobs/5119674733#step:11:15) is strange: "Permission denied" when importing the c-extension. Is there a compilation problem?

r-devulap commented:
> The [cygwin failure](https://github.com/numpy/numpy/actions/runs/3148780125/jobs/5119674733#step:11:15) is strange: "Permission denied" when importing the c-extension. Is there a compilation problem?

Doesn't look like it from logs. It was able to build and install fine. I am very confused with all the errors. Still working my way through them.

Commit:
BUG: Ensure long/longlong is 8 bytes for 64-bit qsort

DWesl commented:
> The [cygwin failure](https://github.com/numpy/numpy/actions/runs/3148780125/jobs/5119674733#step:11:15) is strange: "Permission denied" when importing the c-extension. Is there a compilation problem?

Very odd: [the error message says it's `numpy.core._multiarray_umath` that had trouble importing](https://github.com/numpy/numpy/actions/runs/3148780125/jobs/5119674733#step:11:14), but [the permissions on the file providing that module](https://github.com/numpy/numpy/actions/runs/3148780125/jobs/5119674733#step:13:42) indicate both read and execute permissions for all users, which is what is required for importing.  [Last time something failed to import despite good permissions](https://github.com/numpy/numpy/pull/18330#issuecomment-814955188), the problem was DLL dependencies not being on PATH, but [I have a check for that now](https://github.com/numpy/numpy/actions/runs/3148780125/workflow#L79) and [it doesn't report missing dependencies](https://github.com/numpy/numpy/actions/runs/3148780125/jobs/5119674733#step:13:79), so I'm confused about what's going on.

Next commit [imports fine before hitting a segfault](https://github.com/numpy/numpy/actions/runs/3160661496/jobs/5145373808#step:11:2728) so the permissions failure seems to have gone away.  Hopefully it stays that way.

r-devulap commented:
> The segfault in SIMD reminds me of [this problem](https://github.com/numpy/numpy/pull/18330#issuecomment-820767278), which was fixed in #18795; maybe that came up again?

That's possible. What would be the easiest way to reproduce this error locally? Any pointers would be helpful :)

DWesl commented:
> > The segfault in SIMD reminds me of [this problem](https://github.com/numpy/numpy/pull/18330#issuecomment-820767278), which was fixed in #18795; maybe that came up again?
> 
> That's possible. What would be the easiest way to reproduce this error locally? Any pointers would be helpful :)

This will mostly be following the [CI workflow](https://github.com/numpy/numpy/blob/main/.github/workflows/cygwin.yml).  On a 64-bit Windows machine (it's possible to adapt the directons for 32-bit windows, but [not recommended](https://cygwin.com/#support)):
- Get [setup-86_64.exe from cygwin.com](https://cygwin.com/setup-x86_64.exe)
- Run the installer: when you get to the "Select Packages" page, install (this may be simpler if you change the view in the package picker from "Categories" to "Full" or "Not Installed":
  - python38-devel 
  - python38-zipp 
  - python38-importlib-metadata
  - python38-cython 
  - python38-pip 
  - python38-wheel 
  - python38-cffi
  - python38-pytz 
  - python38-setuptools 
  - python38-pytest
  - python38-hypothesis 
  - liblapack-devel
  - gcc-fortran 
  - gcc-g++ 
  - git 
  - dash
  - Other packages you may find useful for debugging:
    - python38-debuginfo
    - cygwin-debuginfo
    - gdb
  - libopenblas: Might speed up the linear algebra tests, but I don't think you're looking at those
- Start Cygwin: by default it will give you a start menu item and a desktop shortcut, but you can also open a command prompt, navigate to the directory where you installed Cygwin (`C:\cygwin64` by default) and run `Cygwin.bat`.  You can run `mintty -` to get a different terminal, or just deal with blocky fonts and stay in the command prompt.
- Set up development environment according to the [build](https://numpy.org/doc/stable/user/building.html#building-from-source) and [test](https://numpy.org/doc/stable/dev/development_environment.html) guidelines
  - `python3.8 -m pip install 'setuptools<49.2.0' pytest pytz cffi pickle5 importlib_metadata typing_extensions`
  - `python3.8 -m pip install -r test_requirements.txt`
- Build NumPy (`runtests.py` should handle this fine, but the long version is less likely to run into [fork failures](https://www.cygwin.com/faq.html#faq.using.fixing-fork-failures))
  - `python3.8 setup.py bdist_wheel` (The Cygwin CI job [uploads the wheel file on failure](https://github.com/numpy/numpy/blob/26d690e5abe9dc7eb74deb807ea2c4842aa91c70/.github/workflows/cygwin.yml#L71-L76), so you can download that and skip this step if you want)
  - `python3.8 -m pip install dist/numpy-*cp38*.whl` (wheel file name will change with every patch update to Cygwin, so you may need to delete old wheels every so often)
  - `dash "tools/rebase_installed_dlls_cygwin.sh" 3.8`
  - `python3.8 runtests.py -n -vv`
- Debug failures using whatever means you normally use.  You might find it useful to install `pytest-forked` and run just the tests that are likely to fail with `python3.8 runtests.py -n -- --forked`, but forking before every test is too slow on Cygwin to be useful for more than a couple dozen tests.

Another option would be to mark the new functions `NPY_FINLINE` one at a time until the segfault goes away or moves.

[This page](http://www.cygwin.net/cygwin/cygwin-ug-net/using-cygwinenv.html#cygwinenv-implemented-options) has a brief description of how to use the `CYGWIN` environment variable to make programs (including the python process running the numpy tests) start `gdb` on segfault: I don't know how to set environment variables on WIndows, so I'm not sure how useful that is (some Cygwin options need to be set before any Cygwin process starts, but `env CYGWIN='error_start:C:\cygwin64\bin\gdb.exe' python3.8 runtests.py -n -vv` might still work)

r-devulap commented:
@DWesl Thanks for that! And yup, its the same problem as you mentioned! Bunch of aligned mov instructions to unaligned stack pointer. 

```
0x5af801157 <sort_128_32bit<vector<float>, float>(float*, int32_t)+4279>        vmovaps %zmm1,0x120(%rsp)
0x5af801162 <sort_128_32bit<vector<float>, float>(float*, int32_t)+4290>        vzeroupper
0x5af801165 <sort_128_32bit<vector<float>, float>(float*, int32_t)+4293>        call   0x5af7f12b0 <sort_zmm_32bit<vector<float>, __vector(16) float>(float __attribute__ ((vector_size(16))))
0x5af80116a <sort_128_32bit<vector<float>, float>(float*, int32_t)+4298>        vmovups 0x20(%rsp),%zmm3
0x5af801175 <sort_128_32bit<vector<float>, float>(float*, int32_t)+4309>        mov    %r13,%rdx
0x5af801178 <sort_128_32bit<vector<float>, float>(float*, int32_t)+4312>        mov    %r12,%rcx
0x5af80117b <sort_128_32bit<vector<float>, float>(float*, int32_t)+4315>        vmovaps 0x160(%rsp),%zmm0
0x5af801186 <sort_128_32bit<vector<float>, float>(float*, int32_t)+4326>        vmovaps %zmm3,0x120(%rsp)
0x5af801191 <sort_128_32bit<vector<float>, float>(float*, int32_t)+4337>        vmovups %zmm0,0xa0(%rsp)
0x5af80119c <sort_128_32bit<vector<float>, float>(float*, int32_t)+4348>        vzeroupper
```

r-devulap committed:
MAINT: Force inline bitonic network functions

Commit:
ENH: Use npyv_* for missing intrinsics in gcc-6

r-devulap commented:
I'm trying to reproduce the Windows Python38-32bit-fast CI failure (followed the steps listed in `azure-steps-windows.yml`). I can't reproduce the failing tests but I see a different test fail: 

```
FAILED numpy\distutils\tests\test_mingw32ccompiler.py::test_build_import - ValueError: 'nm.exe' found but it does not support 64-bit dlls when using 64-bit python. Supported formats: 'b'supported targets: pe-i386 pei-i386 elf32-i386 e...
1 failed, 22545 passed, 545 skipped, 1306 deselected, 35 xfailed, 28 xpassed in 703.17s (0:11:43)
```

Any idea where I am going wrong?

mattip commented:
> `it does not support 64-bit dlls when using 64-bit python`

Somehow you are running [the test](https://github.com/numpy/numpy/blob/3cf2ca1611ecaf1d13c41075a7025b96e54e3ee3/numpy/distutils/tests/test_mingw32ccompiler.py#L22) with a 64-bit python (`sys.maxsize` is used to check)

r-devulap commented:
Yes looks like it. Doesn't the CI use the python.exe downloaded from here:? https://github.com/numpy/numpy/blob/3cf2ca1611ecaf1d13c41075a7025b96e54e3ee3/azure-steps-windows.yml#L11  

And that is a 64-bit one: 

```
PS C:\Users\rdevulap\numpy> ..\pypy3\python.exe --version
Python 3.8.12 (0089b4a7ab2306925a251b35912885d52ead1aba, Mar 16 2022, 13:51:04)
[PyPy 7.3.9 with MSC v.1929 64 bit (AMD64)]
```

mattip commented:
No, that is PyPy-only. Otherwise [`$PYTHON_VERSION` and $PYTHON_ARCH](https://github.com/numpy/numpy/blob/3cf2ca1611ecaf1d13c41075a7025b96e54e3ee3/azure-pipelines.yml#L259) are used in the [UsePythonVersion](https://github.com/numpy/numpy/blob/3cf2ca1611ecaf1d13c41075a7025b96e54e3ee3/azure-steps-windows.yml#L2) task.

r-devulap commented:
Might need some help reproducing and figuring out why the tests fail on WIN32 and MacOS. Disabling it for now. Will open an issue so we keep track of it. I am guessing the fix should be simple, but I am somehow struggling to reproduce the errors.

Commit:
MAINT: Disable AVX-512 qsort on macOS and WIN32

Commit:
BUG: Do not use a global static const __m512 variable

Commit:
ENH: Add AVX-512 based 16-bit dtype sort

This reverts commit 138ba7583253e7540a206e7f0df3edcd5e26c518.

Commit:
MAINT: Fix comment

Commit:
MAINT: Use loadu intrinsic instead of set1_epi16

gcc-8 is missing the _mm512_set1_epi16 intrinsic

r-devulap commented:
The ever elusive "All checks have passed" 🤦‍♂️But I this last [failure  ](https://app.travis-ci.com/github/numpy/numpy/jobs/585480747) has nothing to do with this patch.

Commit:
TST: Add quicksort test coverage for all 16, 32, 64 bit dtypes

Commit:
MAINT: Fix linter errors

r-devulap commented:
ping .. any idea why the two failures? one is on Arm64 and other one might also not be related (this patch is disabled on 32-bit windows).

mattip commented:
The aarch64 run is timing out at ~80%

Commit:
ENH: Use AVX-512 qsort for half precision float

r-devulap committed:
BENCH: Add float16 to sort benchmarks

r-devulap committed:
TST: Add test for float16 quicksort

r-devulap commented:
Added AX-512 quicksort for float16. This was the tricky one and I have a feeling it can be improved further. Benchmarks: 

```
       before           after         ratio
     [db7414b7]       [cf13cd63]
     <main>           <avxsort>
-       196±0.4μs        140±0.1μs     0.72  bench_function_base.Sort.time_sort('quick', 'float16', ('ordered',))
-         462±3μs       131±0.05μs     0.28  bench_function_base.Sort.time_sort('quick', 'float16', ('sorted_block', 1000))
-         609±2μs        129±0.1μs     0.21  bench_function_base.Sort.time_sort('quick', 'float16', ('sorted_block', 10))
-         590±7μs        124±0.1μs     0.21  bench_function_base.Sort.time_sort('quick', 'float16', ('sorted_block', 100))
-        698±10μs        129±0.1μs     0.18  bench_function_base.Sort.time_sort('quick', 'float16', ('random',))
-       174±0.4μs      6.58±0.03μs     0.04  bench_function_base.Sort.time_sort('quick', 'float16', ('uniform',))
-       175±0.2μs      6.58±0.06μs     0.04  bench_function_base.Sort.time_sort('quick', 'float16', ('reversed',))
```

r-devulap committed:
Fix linter errors'

Commit:
BUG: Use log2 instead a builtin

r-devulap commented:
@mattip The ARM64 failure on TravisCI is unlikely to be related to this patch (this patch is x86 code only). Looks like other PR's have the similar failures too: https://app.travis-ci.com/github/numpy/numpy/jobs/587139824 and https://app.travis-ci.com/github/numpy/numpy/jobs/587353553

mattip commented:
@seiko2plus any thoughts?

r-devulap commented:
just a friendly ping :)

seiko2plus commented:
@r-devulap, sorry, currently I suffer from lack of time, please just give me 2 or 3 days max to give a full review.

r-devulap commented:
@seiko2plus Didnt mean to rush you. Please take your time.

seiko2plus reviewed (changes requested):
Sorry for taking forever. Could you try to simplify the code and avoid complexity? 
It will make it easy for me to review the actual SIMD kernels and enhance or maintain them later.

I suggest the following modifications:

- Move all AVX512 utilities into header `avx512-common-qsort.h` or `avx512_common.hpp`; packing them will shrink code
  and avoid forward declarations.
- The primary implementation should move into the dispatchable sources rather than creating new headers.
- Use fixed-width data types
- Count on function overloading and avoid defining static class functions.
- Let the compiler deduce your types, don't enforce it.
- Count on template traits over fixed-width data types and discard the current tags (`*_tags`).
- Make it look and feel like C++ rather than C mixed with templates. We're currently discussing
  adopting the Google C++ style guide.
- Avoid using the #definitions of CPU groups names and use the #definitions of CPU features instead;
  It makes it easier to understand the required features; also, AMD is on track now.

Here is also a complete example that is putting the above
points into action(I didn't test it but it should deliver the idea):

<details>
<summary> Utility header "avx512_common.hpp" </summary>

```C++
#ifndef NUMPY_SRC_COMMON_NPYSORT_AVX512_COMMON_HPP_
#define NUMPY_SRC_COMMON_NPYSORT_AVX512_COMMON_HPP_

#include <cstdint>

#include "simd/simd.h"

// define a seprated namespace for the x86 simd
namespace np { namespace simd_avx512 {

using std::int32_t;
using std::int64_t;


template<typename Tlane>
struct SimdTraits;

template<> struct SimdTraits<int32_t> {
    using VecType = __m512i;
    using MaskType = __mmask16;
};

template<> struct SimdTraits<int64_t> {
    using VecType = __m512i;
    using MaskType = __mmask8;
};

template<> struct SimdTraits<float> {
    using VecType = __m512;
    using MaskType = __mmask16;
};

template<> struct SimdTraits<double> {
    using VecType = __m512d;
    using MaskType = __mmask8;
};

template<typename Tlane>
class Vec final {
    using VecType = typename SimdTraits<Tlane>::VecType;
    VecType reg_;
 public:
    /// Construct from SIMD register type.
    Vec(VecType r) : reg_(r) {}
    /// Type cast operator to convert to SIMD register type
    operator VecType() const
    { return reg_; }
};
template<typename Tlane>
class Mask final {
    using MaskType = typename SimdTraits<Tlane>::MaskType;
    MaskType reg_;
 public:
    /// Construct from SIMD register type.
    Mask(MaskType r) : reg_(r) {}
    /// Type cast operator to convert to SIMD register type
    operator MaskType() const
    { return reg_; }
};

/// Small functions can implmented within single MACRO
#define IMPL_SORT_X86_AVX512_UTILITIES(TLANE, SFX, USFX, CAST)  \
    NPY_FINLINE Vec<TLANE> Set(TLANE v)                         \
    { return _mm512_set1_##SFX(v); }                            \
    NPY_FINLINE Vec<TLANE> Max(const Vec<TLANE> &a,             \
                               const Vec<TLANE> &b)             \
    { return _mm512_max_##SFX(a, b); }                          \
    NPY_FINLINE Vec<TLANE> Min(const Vec<TLANE> &a,             \
                               const Vec<TLANE> &b)             \
    { return _mm512_min_##SFX(a, b); }                          \
    NPY_FINLINE Vec<TLANE> Move(const Vec<TLANE> &x,            \
                                const Mask<TLANE> &mask,        \
                                const Vec<TLANE> &y)            \
    { return _mm512_mask_mov_##SFX(x, mask, y); }               \
    NPY_FINLINE void CompressStoreu(TLANE *mem,                 \
                               const Mask<TLANE> &mask,         \
                               const Vec<TLANE> &x)             \
    {                                                           \
        return _mm512_mask_compressstoreu_##SFX(                \
                       (CAST*)mem, mask, x);                    \
    }

IMPL_SORT_X86_AVX512_UTILITIES(int32_t, epi32, epi32, void)
IMPL_SORT_X86_AVX512_UTILITIES(int64_t, epi64, epi32, void)
IMPL_SORT_X86_AVX512_UTILITIES(float, ps, ps, float)
IMPL_SORT_X86_AVX512_UTILITIES(double, pd, pd, double)
#undef IMPL_SORT_X86_AVX512_UTILITIES

// TODO: Wraps the rest intrinsics

template<typename Tlane>
NPY_FINLINE void COEX(Vec<Tlane> &a, Vec<Tlane> &b)
{
    auto temp = a;
    a = Min(a, b);
    b = Max(temp, b);
}

template<typename Tlane>
NPY_FINLINE Vec<Tlane> CmpMerge(const Vec<Tlane> &in1, const Vec<Tlane> &in2,
                                const Mask<Tlane> &mask)
{
    auto min = Min(in2, in1);
    auto max = Max(in2, in1);
    return Move(min, mask, max); // 0 -> min, 1 -> max
}

/*
 * Parition one ZMM register based on the pivot and returns the index of the
 * last element that is less than equal to the pivot.
 */
template <typename Tlane>
NPY_FINLINE int32_t PartitionVec(Tlane *arr,
                                 int64_t left,
                                 int64_t right,
                                 const Vec<Tlane> &curr_vec,
                                 const Vec<Tlane> &pivot_vec,
                                 Vec<Tlane> &smallest_vec,
                                 Vec<Tlane> &biggest_vec)
{
    // which elements are larger than the pivot
    Mask<Tlane> gt_mask = Ge(curr_vec, pivot_vec);
    int32_t amount_gt_pivot = _mm_popcnt_u32((int32_t)gt_mask);

    CompressStoreu(arr + left, Not(gt_mask), curr_vec);
    CompressStoreu(arr + right - amount_gt_pivot, gt_mask, curr_vec);

    smallest_vec = Min(curr_vec, *smallest_vec);
    biggest_vec = Max(curr_vec, *biggest_vec);
    return amount_gt_pivot;
}
}} // namespace np::simd_avx512
#endif // NUMPY_SRC_COMMON_NPYSORT_AVX512_COMMON_HPP_
```
</details>

<details>
<summary>32-bit sorting "avx512_qsort32.dispatch.cpp" </summary>

```C++

#include "avx512_qsort.hpp" // hold the declarations of the exported symbols
#include "avx512_common.hpp"

namespace np { namespace simd_avx512 {
// extra helper limited to 32-bit sort operations
/*
 * Assumes zmm is random and performs a full sorting network defined in
 * https://en.wikipedia.org/wiki/Bitonic_sorter#/media/File:BitonicSort.svg
 */
template <typename Tlane>
NPY_FINLINE Vec<Tlane> BitonicSort(const Vec<Tlane> &v)
{
    v = CmpMerge(v, Shuffle<2, 3, 0, 1>(v), Mask<Tlane>(0xAAAA));
    v = CmpMerge(v, Shuffle<0, 1, 2, 3>(v), Mask<Tlane>(0xCCCC));
    // ...
    return v;
}

template <typename Tlane>
NPY_FINLINE Vec<Tlane> BitonicMerge(const Vec<Tlane> &v)
{
    //...
}
// the rest of the helper functions

template <typename Tlane>
NPY_FINLINE void QSortImpl(Tlane *arr, int64_t left,
                          int64_t right, int64_t max_iters)
{
    // ...
}

void NPY_CPU_DISPATCH_CURFX(QSort)(int32_t *arr, intptr_t arrsize)
{
    QSortImpl(arr, 0, arrsize - 1, 2 * (int64_t)log2(arrsize));
}

void NPY_CPU_DISPATCH_CURFX(QSort)(int32_t *arr, intptr_t arrsize)
{
    // QsortImpl(...)
}

void NPY_CPU_DISPATCH_CURFX(QSort)(float *arr, intptr_t arrsize)
{
    // QsortImpl(...)
}

}} // namespace np::simd_avx512

```
</details>

same goes for 16-bit & 64-bit data types

<details>
<summary> Forward declarations of the exported symbols "avx512_qsort.hpp" </summary>

```C++
#ifndef NUMPY_SRC_COMMON_NPYSORT_AVX512_QSORT_HPP_
#define NUMPY_SRC_COMMON_NPYSORT_AVX512_QSORT_HPP_
#include <cstdint>

namespace np { namespace simd_avx512 {
// maybe you should add new header e.g common/npstd.hpp
// and implict all std types under namespace np
using std::intptr_t;
using std::int32_t;
// etc

#ifndef NPY_DISABLE_OPTIMIZATION
#include "avx512_qsort32.dispatch.h"
#endif

NPY_CPU_DISPATCH_DECLARE(void QSort, (uint32_t *start, intptr_t arrsize))
NPY_CPU_DISPATCH_DECLARE(void QSort, (int32_t  *start, intptr_t arrsize))
NPY_CPU_DISPATCH_DECLARE(void QSort, (float    *start, intptr_t arrsize))
// the rest

#ifndef NPY_DISABLE_OPTIMIZATION
#include "avx512_qsort16.dispatch.h"
#endif
NPY_CPU_DISPATCH_DECLARE(void QSort, (uint16_t *start, intptr_t arrsize))
NPY_CPU_DISPATCH_DECLARE(void QSort, (int16_t  *start, intptr_t arrsize))
// assume wraps uint16_t within an struct, we need a decent class for it
// I will work on it soon.
NPY_CPU_DISPATCH_DECLARE(void QSort, (Half     *start, intptr_t arrsize))


}} // namespace np::simd_avx512
#endif // NUMPY_SRC_COMMON_NPYSORT_AVX512_QSORT_HPP_

```
</details>


<details>
<summary>The runtime dispatch</summary>

```C++

#include "avx512_qsort.hpp"

namespace qsort_meta {
    template<int size, bool unsig>
    struct IntBySize;
    template<bool unsig>
    struct IntBySize<sizeof(uint16_t), unsig> {
        using Type = typename std::conditional<
            unsig, uint16_t, int16_t>::type;
    };
    template<bool unsig>
    struct IntBySize<sizeof(uint32_t), unsig> {
        using Type = typename std::conditional<
            unsig, uint32_t, int32_t>::type;
    };
    template<bool unsig>
    struct IntBySize<sizeof(uint64_t), unsig> {
        using Type = typename std::conditional<
            unsig, uint64_t, int64_t>::type;
    };

    template<typename T>
    struct FixedWidth {
        using Type = typename std::conditional<
            std::is_integral<T>::value,
            typename IntBySize<sizeof(T), std::is_unsigned<T>::value>::Type,
            T
        >::type;
    };
} // namespace qsort_meta


template<typename T, typename TF=typename qsort_meta::FixedWidth<T>::Type>
static bool quicksort_dispatch(T *start, npy_intp num)
{
    void (*dispfunc)(TF*, intptr_t) = nullptr;
    if (sizeof(T) == sizeof(uint16_t)) {
        #ifndef NPY_DISABLE_OPTIMIZATION
            #include "avx512_qsort16.dispatch.h"
        #endif
        NPY_CPU_DISPATCH_CALL_XB(dispfunc = np::simd_avx512::template QSort<TF,>);
    }
    else if (sizeof(T) == sizeof(uint32_t)) {
        #ifndef NPY_DISABLE_OPTIMIZATION
            #include "avx512_qsort32.dispatch.h"
        #endif
        NPY_CPU_DISPATCH_CALL_XB(dispfunc = np::simd_avx512::template QSort<TF>);
    }
    else if (sizeof(T) == sizeof(uint64_t)) {
        #ifndef NPY_DISABLE_OPTIMIZATION
            #include "avx512_qsort64.dispatch.h"
        #endif
        NPY_CPU_DISPATCH_CALL_XB(dispfunc = np::simd_avx512::template QSort<TF>);
    }
    if (dispfunc) {
        (*dispfunc)(reinterpret_cast<TF*>(start), static_cast<intptr_t>(num));
        return true;
    }
    return false;
}
```
</details>

r-devulap commented:
Apologies for the late response, was out on a long holiday with no access to my computer.

It seems like there are other open source libraries that have expressed interest in using the contents of this PR. I would ideally like to avoid maintaining multiple forks of this project and to that end I created a new repository with the same license as NumPy here: https://github.com/intel/x86-simd-sort. Would NumPy maintainers be open to using this repository as git submodule so that we have just copy of it? I will be the maintainer of this project and we are working on adding more functions in the future (example vectorized `argsort`). If that doesn't work, I am open to hearing alternative solutions.

mattip commented:
Let's discuss this at the next [triage meeting](https://hackmd.io/68i_JvOYQfy9ERiHgXMPvg?view)

seiko2plus commented:
@r-devulap, as we discussed in the triage meeting, it's OK to accept it as a submodule.

Commit:
Adding x86-simd-sort as submodule

Commit:
Add x86 simd sort dispatch files to meson.build

Commit:
Fetch submodules in macOS and Windows build

Commit:
Update to latest commit x86-simd-sort

r-devulap commented:
How do I add submodules in Windows CI job? Adding `submodules: true `to this line work? https://github.com/numpy/numpy/blob/d352270849b3cbff4fcf5511551a247065474f01/azure-pipelines.yml#L290

mattip commented:
The pipeline [could not run](https://dev.azure.com/numpy/numpy/_build/results?buildId=28058&view=results):
```
/azure-pipelines.yml (Line: 293, Col: 7): Unexpected value 'submodules'
```

r-devulap commented:
> Can you call `git submodules update --init` ?

like this? 

```
    steps:
    - template: azure-steps-windows.yml
    - script: git submodule update --init
      displayName: 'Fetch submodules'
```

Commit:
Fix azure-pipelines.yml to checkout submodules

r-devulap commented:
Looks like that worked, `azure-steps-windows.yml` was the right file to edit.

r-devulap commented:
All tests pass. Modified PR to use `x86-simd-sort` as a [submodule](https://github.com/intel/x86-simd-sort).

seiko2plus commented:
@r-devulap, Would you please take a look at examples within this comment https://github.com/numpy/numpy/pull/22315#pullrequestreview-1203224025? suppose to provide a clean way for managing the runtime dispatching.

r-devulap commented:
> @r-devulap, Would you please take a look at examples within this comment [#22315 (review)](https://github.com/numpy/numpy/pull/22315#pullrequestreview-1203224025)? suppose to provide a clean way for managing the runtime dispatching.

Looking into it.

r-devulap commented:
> @r-devulap, Would you please take a look at examples within this comment [#22315 (review)](https://github.com/numpy/numpy/pull/22315#pullrequestreview-1203224025)? suppose to provide a clean way for managing the runtime dispatching.

I understand that the issue with my runtime dispatch is that it looks like C code mixed with templates (by explicitly spelling the various typenames). My C++ knowledge isnt the best, so please help me understand your approach. You are helping the compiler deduce the types based on overloaded functions like these: 

```
NPY_CPU_DISPATCH_DECLARE(void QSort, (uint32_t *start, intptr_t arrsize))
NPY_CPU_DISPATCH_DECLARE(void QSort, (int32_t  *start, intptr_t arrsize))
NPY_CPU_DISPATCH_DECLARE(void QSort, (float    *start, intptr_t arrsize))
```

But during runtime dispatch you call the QSort function using a template argument: 

`NPY_CPU_DISPATCH_CALL_XB(dispfunc = np::simd_avx512::template QSort<TF>);
`

But `Qsort ` isn't a template function. How do I get those to work together?

r-devulap commented:
I am also not sure how to have all the functions in one dispatch source file with multiple targets: `avx512_skx, avx512_icl`. Vectorized quicksort for 16-bit data works only on ICL (and above) and NPY_CPU_DISPATCH_CURFX and NPY_CPU_DISPATCH_CALL_XB shouldn't generate a function for `avx512_skx ` target. I couldn't get the dispatcher to work with this constraint and so I them in two separate dispatch files with their own targets.

seiko2plus committed:
ENH: Towards modern C++

 This patch initializes new C++ headers and also brings new
 namespace `np::` to break away from the current approach
 of using C++ which tends not to be drawn into modernity.

seiko2plus committed:
ENH, SIMD: reimplement CPU dispatching of qsort

  For a Few C++ More

seiko2plus committed:
ENH, SIMD: include npy_cpu_dipatch.h by npy_config.h

  To guarantee of having #defs NPY_HAVE_[CPU features] in the scope

seiko2plus committed:
ENH, SIMD: removes #NPY_ENABLE_AVX512_QSORT and use #directives instead

seiko2plus committed:
fix up meson

seiko2plus committed:
fix up up meson

seiko2plus committed:
No need for add x86-simd-sort as global directory

r-devulap commented:
@seiko2plus I appreciate your help, but please don't waste your time writing code for my patch. All I needed was a hint to write the declaration syntax: 

`NPY_CPU_DISPATCH_DECLARE(template <typename T> void QSort, (T *arr, intptr_t size))`. 

And I am guessing we need to keep the 16-bit qsort dispatch in a separate file and there is no way around it?

seiko2plus commented:
> I appreciate your help, but please don't waste your time writing code for my patch. All I needed was a hint to write the declaration syntax:

> I understand that the issue with my runtime dispatch is that it looks like C code mixed with templates (by explicitly spelling the various typenames). My C++ knowledge isnt the best, so please help me understand your approach. You are helping the compiler deduce the types based on overloaded functions like these:

I wasn't clear enough with this example, I hadn't tested and it was meant to be just a guide.

> I am also not sure how to have all the functions in one dispatch source file with multiple targets: avx512_skx, avx512_icl. Vectorized quicksort for 16-bit data works only on ICL (and above) and NPY_CPU_DISPATCH_CURFX and NPY_CPU_DISPATCH_CALL_XB shouldn't generate a function for avx512_skx target. I couldn't get the dispatcher to work with this constraint and so I put them in two separate dispatch files with their own targets.

It can be done if we moved the fallback implementation  `aquicksort_(type *vv, npy_intp *tosort, npy_intp num)` to the dispatch-able-source and then add a baseline target for it.

> I appreciate your help, but please don't waste your time writing code for my patch. All I needed was a hint to write the declaration syntax

The purpose of writing this code is to guarantee to correct the current path of the current C++, without adding extra burden on you or entering into multiple rounds of reviews, so frankly speaking I'm saving time rather than wasting it.

seiko2plus commented:
Cygwin test fails:

```Bash
Fatal Python error: Segmentation fault

Current thread 0x0000000a00000010 (most recent call first):
  File "/usr/local/lib/python3.9/site-packages/numpy/core/fromnumeric.py", line 1030 in sort
  File "/usr/local/lib/python3.9/site-packages/numpy/core/tests/test_multiarray.py", line 9868 in test_sort_float
  File "/usr/lib/python3.9/site-packages/_pytest/python.py", line 183 in pytest_pyfunc_call
  File "/usr/lib/python3.9/site-packages/pluggy/_callers.py", line 39 in _multicall
  File "/usr/lib/python3.9/site-packages/pluggy/_manager.py", line 80 in _hookexec
  File "/usr/lib/python3.9/site-packages/pluggy/_hooks.py", line 265 in __call__
  File "/usr/lib/python3.9/site-packages/_pytest/python.py", line 1641 in runtest
  File "/usr/lib/python3.9/site-packages/_pytest/runner.py", line 162 in pytest_run
```

seiko2plus commented:
It seems the following directive needs to be extended to cover cygwin 
 
https://github.com/numpy/numpy/blob/472a47f8ea9aa9ffe933c15ac4c0c148570b1781/numpy/core/src/npysort/quicksort.cpp#L72-L76

seiko2plus commented on .gitmodules:
```diff
@@ -4,3 +4,6 @@
 	url = https://github.com/numpy/SVML.git
+[submodule "numpy/core/src/npysort/x86-simd-sort"]
+	path = numpy/core/src/npysort/x86-simd-sort
+	url = https://github.com/intel/x86-simd-sort
```
The last hash [7d7591cf5927e83e4a1e7](https://github.com/intel/x86-simd-sort/commit/7d7591cf5927e83e4a1e7c4b6f2c4dc91a97889f) should be used instead to avoid any new changes that may break the build.

r-devulap replied:
It is pointing to the latest hash. 

```
>> git submodule
 36f4c898f2255e0c98eb6949cd67381552d5ffea doc/source/_static/scipy-mathjax (heads/master)
 7d7591cf5927e83e4a1e7c4b6f2c4dc91a97889f numpy/core/src/npysort/x86-simd-sort (remotes/origin/HEAD)
 1c5260a61e7dce6be48073dfa96291edb0a11d79 numpy/core/src/umath/svml (remotes/origin/HEAD)
```

mattip replied:
right, the `.gitmodules` file only holds the url, the actual commit [is stored elsewhere](https://devops.stackexchange.com/questions/5371/where-is-the-git-submodule-revision-actually-stored)

Commit:
Disable on CYGWIN

r-devulap commented:
> It seems the following directive needs to be extended to cover cygwin
> 
> https://github.com/numpy/numpy/blob/472a47f8ea9aa9ffe933c15ac4c0c148570b1781/numpy/core/src/npysort/quicksort.cpp#L72-L76

disabled on Cygwin.

mattip commented:
This has become quite a large PR. If I understand correctly it adds new headers, new ways of dispatching, and a new dependency. Could we split this into two: one PR to enhance the dispatching mechanism, including documentation, and another to add AVX512 float16 sorting? 

The speed increase is really nice. Correct me if I am wrong, I would imagine much of it comes from having a native float16 sort. If that is true, then implementing a float16 sort without casting float16 to float32 would also speed up non-AVX512 architectures.

seiko2plus reviewed (approved):
LGTM, Thank you

mattip commented:
Thanks @r-devulap

seberg commented:
This broke MacOS CI with (I have a PR to fix it by ensuring the submodule is initalized so headers are there):
```
In file included from build/src.macosx-11.7-x86_64-3.9/numpy/core/src/npysort/simd_qsort.dispatch.avx512_skx.cpp:21:
/Users/runner/work/1/s/numpy/core/src/npysort/simd_qsort.dispatch.cpp:11:14: fatal error: 'x86-simd-sort/src/avx512-32bit-qsort.hpp' file not found
    #include "x86-simd-sort/src/avx512-32bit-qsort.hpp"
             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```
I presume we are on hardware there were AVX512 use (or compiling for it), is OK and that submodule initialization is right (see gh-23216).  If not, then some guard seems not narrow enough.

seberg commented:
@r-devulap if you would like some visibility on this, do feel free to add a `performance` release note, it seems like that may be worthwhile.

r-devulap commented:
> This broke MacOS CI with (I have a PR to fix it by ensuring the submodule is initalized so headers are there):

I wonder why the macOS CI failure didnt show up here.  This PR already has a [commit ](https://github.com/numpy/numpy/pull/22315/commits/774edbd8a572067556e9860d5e5c23f73107421a)to update submodules in the macOS. 

> @r-devulap if you would like some visibility on this, do feel free to add a `performance` release note, it seems like that may be worthwhile.

Will do, thanks!

seberg commented:
Ohhh weird, so now the submodule init is there twice :/.   It seemed to have been a bit random.  Maybe a weird caching issue?

mvtec-bergdolll commented:
Doesn't this change break this documented promise:

> Previous to numpy 1.4.0 sorting real and complex arrays containing nan values led to undefined behaviour. In numpy versions >= 1.4.0 nan values are sorted to the end. The extended sort order is:

From the Intel x86-simd-sort repo:

> If you expect your array to contain NANs, please be aware that the these routines do not preserve your NANs as you pass them.

jan-wassenberg commented:
A quick heads-up that the x86-simd-sort code (I downloaded HEAD from Github) may have an issue sorting large arrays.

I noticed incorrect results when integrating this into our [vqsort benchmark](https://github.com/google/highway/pull/1140). Happy to discuss in [HN comments](https://news.ycombinator.com/item?id=34810610) or here.

seberg commented:
I opened a milestoned issue.  We should investigate this (correctness and maybe also NaN sorting stability) before the next release (which is plenty of time).  (Happy if discussion continues here.)

Another thing that I hope might happen is to have contributions to expand so this is not/less AVX512 specific.

jan-wassenberg commented:
Sounds good 👍 

> Another thing that I hope might happen is to have contributions to expand so this is not/less AVX512 specific.

Our [vqsort](https://github.com/google/highway/tree/master/hwy/contrib/sort) works on S-SSE3/SSE4/AVX2/AVX-512, Arm NEON/SVE/SVE2, WASM SIMD and RISC-V V.

The x86-simd-sort work cites previous work by one of the coauthors (Mark Blacher) of vqsort, but does not include its faster sorting network and more robust pivot sampling (note: some enhancements are not mentioned in our [paper](https://arxiv.org/abs/2205.05982) because they came later).

r-devulap commented:
> Doesn't this change break this documented promise:
> 
> > Previous to numpy 1.4.0 sorting real and complex arrays containing nan values led to undefined behaviour. In numpy versions >= 1.4.0 nan values are sorted to the end. The extended sort order is:
> 
> From the Intel x86-simd-sort repo:
> 
> > If you expect your array to contain NANs, please be aware that the these routines do not preserve your NANs as you pass them.

The AVX-512 sorting routines does put all the nan's at the end of the array. What it doesn't do is preserve your NAN's bit by bit. Meaning, if you pass 0xFFFFFFFF, it will replace it with `std::nanf("1")`, which is still a NAN. AFAIK, NumPy doesnt distinguish between the two.

r-devulap commented:
> A quick heads-up that the x86-simd-sort code (I downloaded HEAD from Github) may have an issue sorting large arrays.
> 
> I noticed incorrect results when integrating this into our [vqsort benchmark](https://github.com/google/highway/pull/1140). Happy to discuss in [HN comments](https://news.ycombinator.com/item?id=34810610) or here.

@jan-wassenberg Couldn't figure out the bug you encountered from that link (perhaps I didn't look deep enough). Could you elaborate?

seberg commented:
> which is still a NAN. AFAIK, NumPy doesnt distinguish between the two.

Ah, OK.  I agree that NumPy should not have to worry about preserving the NaN payload, thanks for clarifying.  (There are interesting use-cases for using NaN payload to identify NAs which are different from NaN, but that would be a problem to solve once someone wants to do that, it might be for R which does that IIRC.)

jan-wassenberg commented:
> @jan-wassenberg Couldn't figure out the bug you encountered from that link (perhaps I didn't look deep enough). Could you elaborate?

Sure. If you patch in that pull request, but remove the `algo == Algo::kIntel ||` change, running bench_sort will raise an error complaining that the result does not match expectations:

```
BenchSort: i=0 of 10000 lanes: N1=1 -2146707942 -2146801302 vs. -2146300295 -2146707942
Abort at third_party/highway/hwy/contrib/sort/result-inl.h:125: 32-bit sort is incorrect
```
The problem is that the result doesn't match the expected sort order. But now I see what happened: x86-simd-sort doesn't seem to support reverse sorting, but our test also exercised that. Sorry I didn't notice that before - with only ascending order the test passes and it looks like all is well. I've updated our patch.

r-devulap commented:
@jan-wassenberg Sounds good. Thanks for clarifying!

MilesParker commented:
Curious if these changes might impact order for element subsets that are undetermined (i.e. identical values for fields)?
Asking because I'm trying to sort out some regressions in internal tests.

charris commented:
> if these changes might impact order

Not sure what you are asking here. If you need identical elements to be taken in the same order, stable sorts are what you want, quicksort variants can differ in many ways.
